bigtable/src/main/scala/magnolify/bigtable/BigtableType.scala

```diff
@@ -109,112 +151,53 @@ object BigtableField {
-      v.asReadOnlyByteBuffer().asFloatBuffer().get()
-
-    def toByteString(v: Float): ByteString =
-      writeByteString(java.lang.Float.BYTES, _.putFloat(v))
-  }
-
-  implicit val btDouble = new Primitive[Double] {
-    type ValueT = Double
-
-    def fromByteString(v: ByteString): Double =
-      v.asReadOnlyByteBuffer().asDoubleBuffer().get()
-
-    def toByteString(v: Double): ByteString =
-      writeByteString(java.lang.Double.BYTES, _.putDouble(v))
-  }
-
-  implicit val btShort = new Primitive[Short] {
-    type ValueT = Short
-
-    def fromByteString(v: ByteString): Short =
-      v.asReadOnlyByteBuffer().asShortBuffer().get()
-
-    def toByteString(v: Short): ByteString =
-      writeByteString(java.lang.Float.BYTES, _.putShort(v))
-  }
-
-  implicit val btString = new Primitive[String] {
-    type ValueT = String
-
-    def fromByteString(v: ByteString): String =
-      v.toStringUtf8
-
-    def toByteString(v: String): ByteString =
-      ByteString.copyFromUtf8(v)
+  private def primitive[T](
+    capacity: Int
+  )(f: ByteBuffer => T)(g: (ByteBuffer, T) => Unit): Primitive[T] = new Primitive[T] {
+    override def fromByteString(v: ByteString): T = f(v.asReadOnlyByteBuffer())
+    override def toByteString(v: T): ByteString = {
+      val bb = ByteBuffer.allocate(capacity)
+      g(bb, v)
+      bb.rewind()
+      ByteString.copyFrom(bb)
+    }
   }
 
-  implicit val btBoolean = new Primitive[Boolean] {
-      type ValueT = Boolean
-
-      def fromByteString(v: ByteString): Boolean =
-         v.asReadOnlyByteBuffer().asShortBuffer().get() match {
-           case 0 => false
-           case _ => true
-         }
-
-      def toByteString(v: Boolean): ByteString =
-        writeByteString(java.lang.Float.BYTES, _.putShort(if(v) 1 else 0))
+  implicit val btfInt = primitive[Int](java.lang.Integer.BYTES)(_.getInt)(_.putInt(_))
+  implicit val btfLong = primitive[Long](java.lang.Long.BYTES)(_.getLong)(_.putLong(_))
+  implicit val btfFloat = primitive[Float](java.lang.Float.BYTES)(_.getFloat)(_.putFloat(_))
+  implicit val btfDouble = primitive[Double](java.lang.Double.BYTES)(_.getDouble)(_.putDouble(_))
+  implicit val btfByte = primitive[Byte](java.lang.Byte.BYTES)(_.get)(_.put(_))
+  implicit val btfShort = primitive[Short](java.lang.Short.BYTES)(_.getShort)(_.putShort(_))
+  implicit val btfBoolean = from[Byte](_ == 1)(if (_) 1 else 0)
+
+  implicit val btfByteString = new Primitive[ByteString] {
+    override def fromByteString(v: ByteString): ByteString = v
+    override def toByteString(v: ByteString): ByteString = v
   }
+  implicit val btfByteArray = from[ByteString](_.toByteArray)(ByteString.copyFrom(_))
+  implicit val btfString = from[ByteString](_.toStringUtf8)(ByteString.copyFromUtf8(_))
+
+  implicit def btfOption[A](implicit btf: BigtableField[A]): BigtableField[Option[A]] =
+    new BigtableField[Option[A]] {
+      override def get(columnFamily: String, v: Row, k: String): Option[A] = {
```

Comment:
Not sure how much difference it makes really, linear search for both `$k` and `$k.` vs binary search for one.

Reply:
yeah, 2*log(n) should be better that n, but usually there are not so many fields so probably not critical.